### PR TITLE
Extract partial view

### PR DESCRIPTION
**Pull request by adelf**
Implements #23 
Extract partial view feature. In blade editor, select some code, popupmenu > Refactor > Extract partial view(Ctrl+Alt+P)
Shows dialog with new view name input. Provides completion with existing directory names, like "partials.", "shared.layouts." etc.
Views directory finding way - go up from current file and find "views" or "templates" directory. If couldn't - tries to find one from settings(resources/views, etc.), if no one exists, place new view file to project directory.

### CODE REVIEW

File: src/de/espend/idea/laravel/ui/ExtractPartialViewDialog.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    @Override
+    protected String getDimensionServiceKey() {
+        return "Laravel.ExtractPartialViewDialog";
+    }
+
+    @Nullable
+    @Override
+    protected ValidationInfo doValidate() {
+
+        Pattern pattern = Pattern.compile("^[\\w\\d\\-_]+([\\.\\/][\\w\\d\\-_]+)*$");
```
Comment (Haehnchen) [inline src/de/espend/idea/laravel/ui/ExtractPartialViewDialog.java]: can you add a comment with some examples this expression should match?

File: src/de/espend/idea/laravel/ui/ExtractPartialViewDialog.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+        if(!pattern.matcher(viewNameEditor.getText()).find()) return new ValidationInfo("Wrong view name", viewNameEditor.getComponent());
```
Comment (Haehnchen) [inline src/de/espend/idea/laravel/ui/ExtractPartialViewDialog.java]: please no if one liner

if(!pattern.matcher(viewNameEditor.getText()).find())  {
...
}

File: src/de/espend/idea/laravel/blade/actions/ExtractPartialViewHandler.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    @Nullable
+    private PsiDirectory getViewsDirectory(Project project, PsiFile psiFile)
+    {
+        String basePath = project.getBasePath();
+
+        PsiDirectory directory = psiFile.getContainingDirectory();
+        while(directory != null && !directory.getVirtualFile().getPath().equals(basePath)) {
+            if(directory.getName().equals("views") || directory.getName().equals("templates")) {
```
Comment (Haehnchen) [inline src/de/espend/idea/laravel/blade/actions/ExtractPartialViewHandler.java]: maybe iterate through  `ViewCollector.getPaths` and try to check if directory is inside this path. so we can then support custom project structures as well and do not really on `views` or `template` structure

File: src/de/espend/idea/laravel/blade/actions/ExtractPartialViewHandler.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                }
+
+                final String fileName = viewPathParts[viewPathParts.length - 1] + ".blade.php";
+
+                if(directory.findFile(fileName) != null) {
+                    Messages.showMessageDialog(project, "File already exists", "Info", Messages.getErrorIcon());
+                    return;
+                }
+
+                PsiFile newViewFile = PsiFileFactory.getInstance(project).createFileFromText(fileName,
```
Comment (Haehnchen) [inline src/de/espend/idea/laravel/blade/actions/ExtractPartialViewHandler.java]: one line per parameter:
```
.createFileFromText(
  fileName,
  BladeLanguage.INSTANCE,
  selectedText
);
```

File: src/de/espend/idea/laravel/blade/actions/ExtractPartialViewHandler.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        final String canonizedViewName = viewName.replace('/', '.');
+
+        ApplicationManager.getApplication().runWriteAction(() -> {
+            CommandProcessor.getInstance().executeCommand(editor.getProject(), () -> {
+
+                final String[] viewPathParts = canonizedViewName.split("\\.");
+
+                PsiDirectory directory = targetDirectory;
+
+                for(int i = 0; i < viewPathParts.length - 1; i++) {
```
Comment (Haehnchen) [inline src/de/espend/idea/laravel/blade/actions/ExtractPartialViewHandler.java]: as i know `VfsUtil.createDirectoryIfMissing` supports recursive directory creation

Comment (Haehnchen) [inline src/de/espend/idea/laravel/blade/actions/ExtractPartialViewHandler.java]: as i know `VfsUtil.createDirectoryIfMissing` supports recursive directory creation

File: src/de/espend/idea/laravel/blade/actions/ExtractPartialViewHandler.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                PsiFile newViewFile = PsiFileFactory.getInstance(project).createFileFromText(fileName,
+                        BladeLanguage.INSTANCE,
+                        selectedText);
+                CodeStyleManager.getInstance(project).reformat(newViewFile, false);
+
+                directory.add(newViewFile);
+
+                int selectionStart = editor.getSelectionModel().getSelectionStart();
+                int selectionEnd = editor.getSelectionModel().getSelectionEnd();
+                editor.getDocument().replaceString(selectionStart,
```
Comment (Haehnchen) [inline src/de/espend/idea/laravel/blade/actions/ExtractPartialViewHandler.java]: one line per parameter

File: src/de/espend/idea/laravel/blade/actions/ExtractPartialViewHandler.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    @Nullable
+    private PsiDirectory getViewsDirectory(Project project, PsiFile psiFile)
+    {
+        PsiDirectory firstExistingTemplatePath = null;
+        for(TemplatePath templatePath : ViewCollector.getPaths(project)) {
+            final VirtualFile templateDir = VfsUtil.findRelativeFile(templatePath.getPath(), project.getBaseDir());
+
+            if(templateDir != null) {
+
+                if(psiFile.getVirtualFile().getPath().startsWith(templateDir.getPath()))
```
Comment (Haehnchen) [inline src/de/espend/idea/laravel/blade/actions/ExtractPartialViewHandler.java]: code-style: `{` in if line

File: src/de/espend/idea/laravel/ui/ExtractPartialViewDialog.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+public class ExtractPartialViewDialog extends DialogWrapper {
+
+    private EditorTextField viewNameEditor;
+
+    private JPanel panel;
+
+    // View name should be "header", or "partials.header", or "sh-ar_ed.partials.header",
+    // so, words with '_' and '-' separated with dots
+    private static Pattern viewNamePattern = Pattern.compile("^[\\w\\d\\-_]+([\\.\\/][\\w\\d\\-_]+)*$");
```
Comment (Haehnchen) [inline src/de/espend/idea/laravel/ui/ExtractPartialViewDialog.java]: this there i way to support namespace also: `de.espend.idea.laravel.blade.util.BladeTemplateUtil#resolveTemplateName`?

**Comment by adelf** [comment]
I added namespaces support, but user shouldn't write it in new view name textbox. It will be added to @include(' automatically.
I tried VfsUtil.createDirectoryIfMissing but I have to catch IOException, also 2-3 times VirtualFile <-> PsiDirectory casting. Code is not better than current...

**Comment by adelf** [comment]
Seems, namespaces here is useless. Parent and partial view will be in the same namespace. So, there is no need to write namespace in include statement... Let me do some tests...

**Comment by Haehnchen** [comment]
so looks then ready to me. let me know if you need to change something and finally done

**Comment by adelf** [comment]
I'll change it tomorrow

**Comment by adelf** [comment]
I found only little bug with view folder searching. I think it's ok now.

**Pull request merged**